### ISSUE TITLE
infinite loop in TDSCommand.close()

### ISSUE THREAD

**Issue by oferfort** [microsoft/mssql-jdbc#1575]
## Driver version
doesn't reproduce in 6.5.0, reproduced easily in 6.5.1, a bit harder to reproduce in 9.2.1

## SQL Server version
2016

## Client Operating System
Linux/Windows

## JAVA/JVM version
java version "1.8.0_291"
Java(TM) SE Runtime Environment (build 1.8.0_291-b10)
Java HotSpot(TM) 64-Bit Server VM (build 25.291-b10, mixed mode)

## Table schema
CREATE TABLE [dbo].[O_MY_TEST](
	[ID] [nvarchar](56) NOT NULL,
	[CREATED] [datetime] NOT NULL,
	[DATA] [varbinary](max) NULL,
 CONSTRAINT [PK_NC_MY_TEST] PRIMARY KEY NONCLUSTERED 
(
	[ID] ASC,
	[CREATED] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON)
)

## Problem description
Trying to insert records to a table when it is locked. When we release the lock the driver tries to close the last TDSCommand and gets stuck in an infinite loop.

## JDBC trace logs
attached

## Reproduction code

public class PlayGround {

    public static void main(String[] args) throws Exception {
//        ILetsPlay mssqlProducer = new MssqlProducer();
//        mssqlProducer.letsPlay();
        Logger logger = Logger.getLogger("com.microsoft.sqlserver.jdbc");
        logger.setLevel(Level.FINEST);
        Handler handler = new ConsoleHandler();
        handler.setLevel(Level.FINEST);
        handler.setFormatter(new SimpleFormatter());
        logger.addHandler(handler);

        handler = new FileHandler("trace.log");
        handler.setLevel(Level.FINEST);
        handler.setFormatter(new SimpleFormatter());
        logger.addHandler(handler);

        String DB_USER = <REMOVED>;
        String DB_PASSWORD = <REMOVED>;
        String DB_SCHEMA = <REMOVED>;
        int DB_PORT = 1433;
        String DB_IP = <REMOVED>;

        SQLServerConnectionPoolDataSource ds = new SQLServerConnectionPoolDataSource();
        ds.setServerName(DB_IP);
        ds.setDatabaseName(DB_SCHEMA);
        ds.setIntegratedSecurity(false);
        ds.setPortNumber(DB_PORT);
        ds.setUser(DB_USER);
        ds.setPassword(DB_PASSWORD);

        String userPreFix = String.valueOf(UUID.randomUUID());
        int startRowNum = 0 ;
        int stopRowNum = 300000;

        boolean stop = false;
        for (int i = startRowNum; i < stopRowNum && !stop; i++) {
            Connection conn = ds.getConnection();
            PreparedStatement psInsert  = conn.prepareStatement("insert into O_MY_TEST (ID,CREATED,DATA) values (?,?,?) ", Statement.RETURN_GENERATED_KEYS);
            psInsert.setQueryTimeout(3);
            String user = userPreFix + "_" + i;
            Date date =  new Date(System.currentTimeMillis());
            try
            {
                for(int j=0;j<3;j++){
                    byte[] bytes = asBytes(UUID.randomUUID());
                    Blob blob = new javax.sql.rowset.serial.SerialBlob(bytes);
                    psInsert.setString(1,user+"_"+j);
                    psInsert.setDate(2, date);
                    psInsert.setBlob(3, blob);
                    psInsert.addBatch();
                }
                psInsert.executeBatch();
                System.out.println("save query number: "+ i);
                Thread.sleep(50);
            }
            catch (Exception e){
                System.out.println(e.getMessage());
            }
            finally {
                psInsert.close();
                conn.close();
            }
        }
        System.out.println("done");
    }

    public static byte[] asBytes(UUID uuid) {
        ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
        bb.putLong(uuid.getMostSignificantBits());
        bb.putLong(uuid.getLeastSignificantBits());
        return bb.array();
    }

}

When the test application is running execute:
BEGIN TRAN
SELECT * FROM O_MY_TEST WITH(UPDLOCK, HOLDLOCK)
/** wait for 30  seconds. in 9.2.1 you need to lock and unlock several times till the problem reproduces **/
COMMIT TRAN

[trace.log](https://github.com/microsoft/mssql-jdbc/files/6401191/trace.log)

**Comment by rafaelmatiello** [comment]
I have the same problem, glassfish thread in infinite loop with high cpu processing. May be related to network instability when close connection.

When it happens, to resolve, I need to restart the server.

Stack of thread dump:
`at com.microsoft.sqlserver.jdbc.TDSCommand.close(IOBuffer.java:5793)
at com.microsoft.sqlserver.jdbc.SQLServerStatement.discardLastExecutionResults(SQLServerStatement.java:94)
at com.microsoft.sqlserver.jdbc.SQLServerStatement.closeInternal(SQLServerStatement.java:584)
at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.closeInternal(SQLServerPreparedStatement.java:170)
at com.microsoft.sqlserver.jdbc.SQLServerStatement.close(SQLServerStatement.java:596)`

**Comment by shycohen123** [comment]
@rafaelmatiello 
How are you creating your statement? with  Statement.RETURN_GENERATED_KEYS?

**Comment by rafaelmatiello** [comment]
@shycohen123 We use default configuration.
Now, I am updated jdbc to version 9.2.1.jre8 for testing.

**Comment by oferfort** [comment]
> Hello @oferfort,
> 
> The team is looking into this issue now.

Thanks, has there been any progress? The only solution once this problem occurs is to restart the server that is using the driver.
Thanks

**Comment by v-zhangw** [comment]
We were able to replicate the issue on 9.2.1 and will be working on a fix. However we are juggling a lot of work right now and likely will not be able to put time on it until next week.

**Comment by oferfort** [comment]
> We were able to replicate the issue on 9.2.1 and will be working on a fix. However we are juggling a lot of work right now and likely will not be able to put time on it until next week.

Thanks for the update

**Comment by v-zhangw** [comment]
Hello,
It does look like this behaviour was introduced between 6.5.0 and 6.5.1, however that change was made to fix another issue at the time. We do still intend to fix this, but it looks like it will not be a trivial time commitment and we have higher priority work at the moment. This will be backlogged and we will look at it again during issue triage.

**Comment by rafaelmatiello** [comment]
After update driver to version 9.2.1.jre8, the glassfish has been running well for a week.

**Comment by oferfort** [comment]
Thanks @v-zhangw , I'm the triage please remember that once this happens it leaves the server using the driver, useless and you have to restart it

**Comment by oferfort** [comment]
I can update that when we changed the code to use Statement.NO_GENERATED_KEYS the problem doesn't reproduce

**Comment by oferfort** [comment]
hi, any update regarding this issue?

**Comment by lilgreenbird** [comment]
hi @oferfort
This issue is in our backlog it will be considered along with other issues and features when we plan for the next semester

**Comment by alex-rentz** [comment]
Hello, I think I am seeing this same issue. Here is a stack trace of the hanging thread:

```
com.REDACTED.ThreadDetails: prio=4 Id=4427272 RUNNABLE
 at com.microsoft.sqlserver.jdbc.TDSCommand.close(IOBuffer.java:7209)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.discardLastExecutionResults(SQLServerStatement.java:126)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.closeInternal(SQLServerStatement.java:611)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.close(SQLServerStatement.java:623)
 at com.zaxxer.hikari.pool.ProxyStatement.close(ProxyStatement.java:76)
 at org.jooq.impl.ProviderEnabledStatement.close(ProviderEnabledStatement.java:71)
 at org.jooq.tools.jdbc.DefaultStatement.close(DefaultStatement.java:137)
 at org.jooq.tools.jdbc.JDBCUtils.safeClose(JDBCUtils.java:294) ...
 at com.microsoft.sqlserver.jdbc.TDSCommand.close(IOBuffer.java:7209)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.discardLastExecutionResults(SQLServerStatement.java:126)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.closeInternal(SQLServerStatement.java:611)
 at com.microsoft.sqlserver.jdbc.SQLServerStatement.close(SQLServerStatement.java:623)
 at com.zaxxer.hikari.pool.ProxyStatement.close(ProxyStatement.java:76)
 at org.jooq.impl.ProviderEnabledStatement.close(ProviderEnabledStatement.java:71)
 at org.jooq.tools.jdbc.DefaultStatement.close(DefaultStatement.java:137)
 at org.jooq.tools.jdbc.JDBCUtils.safeClose(JDBCUtils.java:294)
 at org.jooq.impl.Tools.safeClose(Tools.java:2016)
 at org.jooq.impl.Tools.safeClose(Tools.java:1997)
 at org.jooq.impl.Tools.safeClose(Tools.java:1990)
 at org.jooq.impl.BatchMultiple.execute(BatchMultiple.java:130)
 at org.jooq.impl.BatchMultiple.execute(BatchMultiple.java:78)
 at ...
```

Proprietary stack trace elements removed. We are seeing high CPU utilization accompanying this thread in the thread dump, which makes me think it is most likely an infinite loop.

We are on 6.1.7.jre8-preview. I am trying to see if we can upgrade. Has this been fixed?

**Comment by lilgreenbird** [comment]
There's been a fix relating to this issue could you please upgrade to the [latest release](https://github.com/microsoft/mssql-jdbc/releases/tag/v10.2.1) and see if it fixes your issue?

**Comment by Jeffery-Wasty** [comment]
Hi @oferfort,

I want to follow up on the previous comment. We've had a fix related to this issue in 10.2.1, does this fix your issue?

**Comment by oferfort** [comment]
we are still on 9.4.1 so don't know yet
